tools/reduce.py

```diff
@@ -0,0 +1,303 @@
+                lambda: logging.info(
+                    f"scip-clang temporary directory persisted at {tempdir}"
+                )
+            )
+            tempdir = pathlib.Path(tempdir)
+            self.entry.directory = tempdir
+
+            shutil.copyfile(absolute_start_tu_path, tempdir.joinpath(start_tu_path))
+            compdb = tempdir.joinpath("compile_commands.json")
+            with open(compdb, "w") as compdb_file:
+                json.dump([self.entry.to_dict()], compdb_file)
+
+            run_sh = tempdir.joinpath("run.sh")
+            with open(run_sh, "w") as run_sh_file:
+                assert self.scip_clang.is_absolute()
+                run_sh_file.write(
+                    textwrap.dedent(
+                        f"""\
+                    #!/usr/bin/env bash
+                    {self.scip_clang} --worker-mode=compdb --compdb-path={compdb} > out.log 2>&1
+                    set -e
+                    grep -E '{self.scip_clang_output_pattern.pattern}' out.log
+                    """
+                    )
+                )
+            run_sh.chmod(run_sh.stat().st_mode | stat.S_IEXEC)
+
+            result = subprocess.run(
+                ["creduce", str(run_sh), str(start_tu_path)] + self.extra_args,
+                cwd=tempdir,
+            )
+            if result.returncode != 0:
+                sys.exit(result.returncode)
+
+    def fix_paths(self):
+        """
+        Adjusts the paths in the argument list so that the compilation database
+        entry can be relocated correctly to a temporary directory.
+
+        Post-condition:
+        - Include paths, such as using -I and -isysroot use absolute paths
+        - The path for the main TU has a relative path
+
+        Creduce creates temporary directories, so we want all the paths to
+        work properly from those directories too.
+        """
+        original_path = self.entry.filepath
+        if self.entry.filepath.is_absolute():
+            self.entry.filepath = self.entry.filepath.relative_to(self.project_root)
+        tu_path_str = str(self.entry.filepath)
+
+        def fix_path(maybe_path: str) -> str:
+            if os.path.isabs(maybe_path):
+                return maybe_path
+            abs_path = self.entry.directory.joinpath(maybe_path)
+            if abs_path.exists():
+                return str(abs_path)
+            # Maybe it wasn't a path argument after all...
+            return maybe_path
+
+        for i, arg in enumerate(self.entry.arguments):
+            if tu_path_str in arg:
+                if os.path.isabs(arg):
+                    self.entry.arguments[i] = tu_path_str
+            else:
+                match = re.match(self.path_flag_pattern, arg)
+                if match is None:
+                    continue
+                if match.group("path") != "":
+                    self.entry.arguments[i] = match.group("flag") + fix_path(
+                        match.group("path")
```

Comment:
Nagging feeling these half a dozen lines could be a list comprehension with regex.sub.

Reply:
This is a nit but you could consume a list to avoid processing an entry twice, something like:

```py
args = self.entry.arguments[:]
rewritten_args = []
while args:
  arg = args.pop(0)
  ...
  if match.group("maybe_path"):
    rewritten_args.push(f'{match.group("flag")}{fix_path(match.group("maybe_path"))}')
  elif args:
    ... mumble args.pop(0) ...
  else:
    ... maybe warn about expecting an argument but finding none; the arg list has been truncated ...
```

Reply:
I think having extra pushing/popping would make this code more complicated, and this code is linear anyways. Initially, I wrote this to do something similar (in terms of avoiding processing twice), where I'd save a boolean and rewrite on the next iteration, but also made the code complicated.

The size of these argument lists, barring pathological cases (like that one customer), should be <1000, so it shouldn't matter. The running time of this code will be dominated by C-reduce, which may take up to minutes to reduce the code.

Reply:
I'm not concerned about efficiency but maintainability in that some arguments are considered twice which is a bit subtle. But this is fine as-is, we can iterate later if we want.